test/DateYearDurationTest.cpp

```diff
@@ -0,0 +1,977 @@
+  // date and time is lost.
+  testLargeYearGYearMonth("2039481726-03", 2039481726, "2039481726-01");
+  testLargeYearGYearMonth("-2039481726-07", -2039481726, "-2039481726-01");
+
+  testLargeYearDate("2039481726-03-01", 2039481726, "2039481726-01-01");
+  testLargeYearDate("-2039481726-02-05", -2039481726, "-2039481726-01-01");
+
+  testLargeYearDatetime("2039481726-01-01T12:00:00", 2039481726,
+                        "2039481726-01-01T00:00:00");
+  testLargeYearDatetime("2039481726-01-01T00:13:00", 2039481726,
+                        "2039481726-01-01T00:00:00");
+  testLargeYearDatetime("-2039481726-01-01T00:00:14", -2039481726,
+                        "-2039481726-01-01T00:00:00");
+}
+
+TEST(Date, parseErrors) {
+  using D = DateYearOrDuration;
+  using E = DateParseException;
+  ASSERT_THROW(D::parseGYear("1994-12"), E);
+  ASSERT_THROW(D::parseGYear("Kartoffelsalat"), E);
+  ASSERT_THROW(D::parseGYearMonth("1994"), E);
+  ASSERT_THROW(D::parseGYearMonth("Kartoffelsalat"), E);
+  ASSERT_THROW(D::parseXsdDate("1994-##-##"), E);
+  ASSERT_THROW(D::parseXsdDate("Kartoffelsalat"), E);
+  ASSERT_THROW(D::parseXsdDatetime("1994-12-13"), E);
+  ASSERT_THROW(D::parseXsdDate("Kartoffelsalat"), E);
+}
+
+TEST(DateYearOrDuration, AssertionFailures) {
+  // These values are out of range.
+  ASSERT_ANY_THROW(DateYearOrDuration(std::numeric_limits<int64_t>::min(),
+                                      DateYearOrDuration::Type::Year));
+  ASSERT_ANY_THROW(DateYearOrDuration(std::numeric_limits<int64_t>::max(),
+                                      DateYearOrDuration::Type::Year));
+
+  // These years have to be stored as a `Date`, not as a large year.
+  ASSERT_ANY_THROW(DateYearOrDuration(-9998, DateYearOrDuration::Type::Year));
+  ASSERT_ANY_THROW(DateYearOrDuration(2021, DateYearOrDuration::Type::Year));
+
+  // Calling `getDate` on an object that is stored as a large year.
+  DateYearOrDuration d{123456, DateYearOrDuration::Type::Year};
+  ASSERT_ANY_THROW(d.getDate());
+}
+
+TEST(DateYearOrDuration, Order) {
+  DateYearOrDuration d1{-12345, DateYearOrDuration::Type::Year};
+  DateYearOrDuration d2{Date{2022, 7, 16}};
+  DateYearOrDuration d3{12345, DateYearOrDuration::Type::Year};
+  DateYearOrDuration d4{
+      DayTimeDuration{DayTimeDuration::Type::Positive, 0, 23, 23, 62.44}};
+  DateYearOrDuration d5{
+      DayTimeDuration{DayTimeDuration::Type::Positive, 1, 24, 23, 62.44}};
+  DateYearOrDuration d6{
+      DayTimeDuration{DayTimeDuration::Type::Negative, 1, 24, 23, 62.44}};
+  DateYearOrDuration d7{
+      DayTimeDuration{DayTimeDuration::Type::Negative, 1, 25, 23, 62.44}};
+
+  ASSERT_EQ(d1, d1);
+  ASSERT_EQ(d2, d2);
+  ASSERT_EQ(d3, d3);
+  ASSERT_EQ(d4, d4);
+  ASSERT_EQ(d5, d5);
+  ASSERT_EQ(d6, d6);
+  ASSERT_EQ(d7, d7);
+  ASSERT_LT(d1, d2);
+  ASSERT_LT(d2, d3);
+  ASSERT_LT(d1, d3);
+  ASSERT_LT(d4, d5);
+  ASSERT_LT(d6, d5);
+  ASSERT_LT(d7, d4);
+  ASSERT_LT(d7, d6);
+}
+
+// SECTION TEST DURATION
```

Comment:
Can you please move them to a separate `...Test.cpp` file , to make things more readable?